ros/cpp/src/ptg_datahub/src/ptg_datahub.cpp

```diff
@@ -311,6 +355,160 @@ class MinimalPublisher : public rclcpp::Node
+        // 32-bit sync
+        // 32-bit length
+        // 32-bit mesh ID
+        // 32-bit vertex count
+        // 32-bit triangle count
+        // Vertex list:
+        //   - 32 bit x
+        //   - 32 bit y
+        //   - 32 bit z
+        // Triangle list:
+        //   - 32 bit index
+
+        // form the spatial mesh message
+        angel_msgs::msg::SpatialMesh message = angel_msgs::msg::SpatialMesh();
+        message.mesh_id = ((frame_data[0] << 24) |
+                           (frame_data[1] << 16) |
+                           (frame_data[2] << 8) |
+                           (frame_data[3]));
+        int vertex_count = ((frame_data[4] << 24) |
+                           (frame_data[5] << 16) |
+                           (frame_data[6] << 8) |
+                           (frame_data[7]));
+        int triangle_count = ((frame_data[8] << 24) |
+                             (frame_data[9] << 16) |
+                             (frame_data[10] << 8) |
+                             (frame_data[11]));
+        triangle_count /= 3;
+
+        // check if this mesh should be removed from the spatial map
+        if ((vertex_count == 0) && (triangle_count == 0))
+        {
+          message.removal = true;
+          message.mesh = shape_msgs::msg::Mesh();
+        }
+        else
+        {
+          message.removal = false;
+          message.mesh = shape_msgs::msg::Mesh();
+          std::vector<shape_msgs::msg::MeshTriangle> triangle_list;
+          std::vector<geometry_msgs::msg::Point> vertex_list;
+
+          int vertex_offset = 12;
+          for (int i = 0; i < (vertex_count * 12); i += 12)
+          {
+            geometry_msgs::msg::Point p = geometry_msgs::msg::Point();
+
+            float x;
+            float y;
+            float z;
+            memcpy(&x, &frame_data[vertex_offset + i], sizeof(x));
+            memcpy(&y, &frame_data[vertex_offset + i + 4], sizeof(y));
+            memcpy(&z, &frame_data[vertex_offset + i + 8], sizeof(z));
+            p.x = x;
+            p.y = y;
+            p.z = z;
+
+            vertex_list.insert(vertex_list.end(), p);
+          }
+
+          int triangle_offset = vertex_offset + 12 * vertex_count;
+          for (int i = 0; i < (triangle_count * 12); i += 12)
+          {
+            shape_msgs::msg::MeshTriangle t = shape_msgs::msg::MeshTriangle();
+            memcpy(&t.vertex_indices[0], &frame_data[triangle_offset + i], sizeof(t.vertex_indices[0]));
+            memcpy(&t.vertex_indices[1], &frame_data[triangle_offset + i + 4], sizeof(t.vertex_indices[1]));
+            memcpy(&t.vertex_indices[2], &frame_data[triangle_offset + i + 8], sizeof(t.vertex_indices[2]));
+            triangle_list.insert(triangle_list.end(), t);
+          }
+
+          message.mesh.vertices = vertex_list;
+          message.mesh.triangles = triangle_list;
+        }
+
+        publisher_->publish(message);
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

Comment:
Maybe a silly question that I didn't think to ask before: why do these `TCPServer*Thread` loops have millisecond sleeps in them?

Reply:
Good question... I thought they were needed at some point but realized they weren't pointless since the `recv` calls are blocking until we get something in the socket right now. I went ahead and removed them.